server/internal/assets/impl.go

```diff
@@ -682,6 +683,148 @@ func (s *Service) ServeOpenAPIv3(ctx context.Context, payload *gen.ServeOpenAPIv
 	}, body, nil
 }
 
+func (s *Service) FetchOpenAPIv3FromURL(ctx context.Context, payload *gen.FetchOpenAPIv3FromURLForm) (*gen.UploadOpenAPIv3Result, error) {
+	authCtx, ok := contextvalues.GetAuthContext(ctx)
+	if !ok || authCtx == nil || authCtx.ProjectID == nil {
+		return nil, oops.C(oops.CodeUnauthorized)
+	}
+
+	// Validate URL
+	parsedURL, err := url.Parse(payload.URL)
+	if err != nil {
+		return nil, oops.E(oops.CodeBadRequest, fmt.Errorf("parse url: %w", err), "invalid URL")
+	}
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		return nil, oops.E(oops.CodeBadRequest, nil, "URL must use http or https scheme")
+	}
+
+	// Fetch the OpenAPI spec from the URL
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, payload.URL, nil)
+	if err != nil {
+		return nil, oops.E(oops.CodeUnexpected, fmt.Errorf("create request: %w", err), "error fetching URL")
+	}
+
+	client := &http.Client{
+		Timeout: 30 * time.Second,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, oops.E(oops.CodeBadRequest, fmt.Errorf("fetch url: %w", err), "error fetching URL")
+	}
+	defer o11y.LogDefer(ctx, s.logger, func() error {
+		return resp.Body.Close()
+	})
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, oops.E(oops.CodeBadRequest, nil, "failed to fetch URL: received status %d", resp.StatusCode)
+	}
+
+	// Determine content type from response or URL
+	contentType := resp.Header.Get("Content-Type")
+	if contentType == "" {
+		// Infer from URL extension
+		ext := strings.ToLower(path.Ext(parsedURL.Path))
+		switch ext {
+		case ".yaml", ".yml":
+			contentType = "application/yaml"
+		case ".json":
+			contentType = "application/json"
+		default:
+			contentType = "application/yaml"
+		}
+	}
+
+	// Parse media type to get just the mime type without parameters
+	mediaType, _, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		mediaType = contentType
+	}
+
```

Comment:
wonder if we should have an allow list if mediaTypes here before we actually download. In case we see something we don't expect.